deepspeed/runtime/zero/stage_1_and_2.py

```diff
@@ -682,7 +696,8 @@ def reduce_gradients(self, pipeline_parallel=False):
         if not self.overlap_comm:
             for i, group in enumerate(self.bit16_groups):
                 for param in group:
-                    if param.grad is not None:
+                    self.point_grad_reduc(param)
+                    if param.grad_reduc is not None:
```

Comment:
I think this could cause an attribute error for non-trainable parameters, e.g., during finetuning. Is this change necessary?

Reply:
The change exists because in Stage 1, Mode 1 `param.reduc_grad` points to `param.grad` and in Mode 2 it points to `param.accum_grad`. In the case of Mode 2 `param.grad` is always None. The `reduc_grad` is used as an alias to be apathetic to the Mode.  I'm not sure about the finetuning process but the array being looped over (`self.bit16_groups`) only contains trainable parameters as far as I'm aware based on the constructor. 
```Python
 298         for i, param_group in enumerate(self.optimizer.param_groups):
 299             partition_id = dist.get_rank(group=self.real_dp_process_group[i])
 300
 301             # push this group to list before modify
 302             # TODO: Explore simplification that avoids the extra book-keeping by pushing the reordered group
 303             trainable_parameters = []
 304             for param in param_group['params']:
 305                 if param.requires_grad:
 306                     param.grad_accum = None
 307                     param.grad_reduc = None
 308                     trainable_parameters.append(param)
 309             self.bit16_groups.append(trainable_parameters)
```

Reply:
Good point, only trainable parameters will be accessed in this loop. Since this loop is only used for stage 1, it means `param.grad` or `param.grad_reduc` is should never be None. Perhaps this should be an assert?